streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -308,12 +305,41 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
         supervisor.start()
         supervisor.awaitTermination()
       }
+
       // Run the dummy Spark job to ensure that all slaves have registered.
       // This avoids all the receivers to be scheduled on the same node.
       if (!ssc.sparkContext.isLocal) {
         ssc.sparkContext.makeRDD(1 to 50, 50).map(x => (x, 1)).reduceByKey(_ + _, 20).collect()
       }
 
+      // Right now, we only honor preferences if all receivers have them
+      val hasLocationPreferences = receivers.map(_.preferredLocation.isDefined).reduce(_ && _)
```

Comment:
Would be nice if this new functionality is put into a function by itself. So that we can unit test it individually to test against different combinations of executors and all, may be different policies in the future.